### PR TITLE
cache: introduce WeightAllocator

### PR DESCRIPTION
**Pull request by CabinfeverB**
Signed-off-by: Cabinfever_B <cabinfeveroier@gmail.com>

### What problem does this PR solve?

Issue Number: ref #5257

### What is changed and how does it work?

```commit-message
add WeightAllocator
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/cache/util.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	for i := 0; i < segNum; i++ {
+		next := segLength
+		if (length % segNum) > i {
+			next++
+		}
+		unitCount += (segNum - i) * next
+		segIndexs = append(segIndexs, next)
+	}
+
+	// If there are 10 results, the weight is [0.143,0.143,0.143,0.143,0.095,0.095,0.095,0.047,0.047,0.047],
```
Comment (rleungx) [inline pkg/cache/util.go:39]: Please update it and move it to the right place.

File: pkg/cache/util.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+type WeightAllocator struct {
+	segIndexs []int
+	wights    []float64
+}
+
+// NewWeightAllocator returns a new WeightAllocator
+func NewWeightAllocator(length, segNum int) *WeightAllocator {
+	segLength := length / segNum
+	segIndexs := make([]int, 0, segNum)
+	wights := make([]float64, 0, length)
```
Comment (rleungx) [inline pkg/cache/util.go:27]:
```suggestion
	weights := make([]float64, 0, length)
```

File: pkg/cache/util.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+// WeightAllocator is used to alloc weight
```
Comment (nolouch) [inline pkg/cache/util.go:17]: Add more comments about this algorithm.

File: pkg/cache/util.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	// and segIndexs is
+	// If there are 3 results, the weight is [0.5,0.33,0.17]
+	unitWeight := 1.0 / float64(unitCount)
+	for i := 0; i < segNum; i++ {
+		for j := 0; j < segIndexs[i]; j++ {
+			wights = append(wights, unitWeight*float64(segNum-i))
+		}
+	}
+	return &WeightAllocator{
+		segIndexs: segIndexs,
```
Comment (bufferflies) [inline pkg/cache/util.go:49]: segIndexs don't need to save.

File: pkg/cache/util.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+// NewWeightAllocator returns a new WeightAllocator
+func NewWeightAllocator(length, segNum int) *WeightAllocator {
+	segLength := length / segNum
+	segIndexs := make([]int, 0, segNum)
+	wights := make([]float64, 0, length)
+
+	unitCount := 0
+	for i := 0; i < segNum; i++ {
+		next := segLength
+		if (length % segNum) > i {
```
Comment (rleungx) [inline pkg/cache/util.go:32]: `length % segNum` is calculated multiple times, is it necessary?

File: pkg/cache/util.go [hunk 6]
```diff
@@ -14,47 +14,52 @@
...
 type WeightAllocator struct {
-	segIndexs []int
-	wights    []float64
+	weights []float64
 }
 
-// NewWeightAllocator returns a new WeightAllocator
+// NewWeightAllocator returns a new WeightAllocator.
 func NewWeightAllocator(length, segNum int) *WeightAllocator {
 	segLength := length / segNum
```
Comment (rleungx) [inline pkg/cache/util.go:29]: Please make sure `segNum` is not equal to 0.

File: pkg/cache/util.go [hunk 7]
```diff
@@ -14,47 +14,52 @@
...
+	segMod := length % segNum
 	segIndexs := make([]int, 0, segNum)
-	wights := make([]float64, 0, length)
+	weights := make([]float64, 0, length)
 
 	unitCount := 0
 	for i := 0; i < segNum; i++ {
 		next := segLength
-		if (length % segNum) > i {
+		if (segMod) > i {
```
Comment (rleungx) [inline pkg/cache/util.go:38]:
```suggestion
		if segMod > i {
```

**Review by rleungx — APPROVED** [review]

File: pkg/cache/util_test.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	wa = NewWeightAllocator(10, 10)
+	for i := 0; i < 10; i++ {
+		re.Equal(1./55.*float64(10-i), wa.Get(i))
+	}
+
+	wa = NewWeightAllocator(10, 1)
+	re.Equal(1./10., wa.Get(0))
+
+	wa = NewWeightAllocator(10, 3)
+	re.Equal(1./21.*3., wa.Get(0))
```
Comment (bufferflies) [inline pkg/cache/util_test.go:39]: You should verify the wa.get(0) is twice thant the wa.get(6), and the mod one is same with the second.

File: pkg/cache/util.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+// If length is 10 and segNum is 3, the weight is [0.143,0.143,0.143,0.143,0.095,0.095,0.095,0.047,0.047,0.047];
+// If length is 10 and segNum is 1, the weight is [0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1];
+// If length is 3 and segNum is 3, the weight is [0.5,0.33,0.17].
+type WeightAllocator struct {
+	weights []float64
+}
+
+// NewWeightAllocator returns a new WeightAllocator.
+func NewWeightAllocator(length, segNum int) *WeightAllocator {
+	if segNum < 1 || length < 0 {
```
Comment (bufferflies) [inline pkg/cache/util.go:32]: can you test (segNum=2 ,length=0) ? I think it maybe wrong.

File: pkg/cache/util.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+package cache
+
+// WeightAllocator is used to alloc weight for cache.
+// When we need to use multiple items from the cache,
+// we may need to assign different weights to these items.
+// WeightAllocator will divide these items into some segments whose number named as segNum which should great than 0.
+// And the items at first segment will be assigned more weight that is `segNum` times that of item at last segment.
+// If you want assign same weights, just input segNum as 1.
+// If length is 10 and segNum is 3, the weight is [0.143,0.143,0.143,0.143,0.095,0.095,0.095,0.047,0.047,0.047];
```
Comment (bufferflies) [inline pkg/cache/util.go:23]: you can describe this sequence  as  arr [3,3,3,3,2,2,2,1,1,1],  and then uniform them :  arr[3,3,3,3,2,2,2,1,1,1]/sum(arr)=arr/21

**Review by bufferflies — COMMENTED** [review]
Rest LGTM， maybe you need to rename the utils file name?

**Comment by CabinfeverB** [comment]
> Rest LGTM， maybe you need to rename the utils file name?

update

**Review by bufferflies — APPROVED** [review]

**Pull request merged**